#1. Import insurance.csv

In [ ]:
import pandas as pd
import numpy as np

# Create a DataFrame
df = pd.read_csv('insurance.csv')


In [ ]:
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

#2. Datapreprocessing

DATAPREPROCESSING: Convert 'sex' to a numerical variable through label encoder

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Initialize the LabelEncoder
labelenc = LabelEncoder()

# Apply LabelEncoder to the 'sex' column
df['sex'] = labelenc.fit_transform(df['sex'])

# Display the updated DataFrame
print(df.head())

   age  sex     bmi  children smoker     region      charges
0   19    0  27.900         0    yes  southwest  16884.92400
1   18    1  33.770         1     no  southeast   1725.55230
2   28    1  33.000         3     no  southeast   4449.46200
3   33    1  22.705         0     no  northwest  21984.47061
4   32    1  28.880         0     no  northwest   3866.85520


OR:

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Original 2D array
targets = np.array(['female', 'male'])

# Initialize the LabelEncoder
labelenc = LabelEncoder()

# Fit the LabelEncoder
labelenc.fit(targets)

# Transform the targets
targets_trans = labelenc.transform(targets)

# Print the results
print("The original data")
print(targets)
print("The transformed data using LabelEncoder")
print(targets_trans)

The original data
['female' 'male']
The transformed data using LabelEncoder
[0 1]


In [ ]:
labelenc.classes_ #basically in alphabetical order

array(['female', 'male'], dtype='<U6')

OR: (rerun cell 1 first)

In [ ]:
# Copy the original 'sex' column
original_sex = df['sex'].copy()

# Initialize the LabelEncoder
labelenc = LabelEncoder()

# Apply LabelEncoder to the 'sex' column
df['sex'] = labelenc.fit_transform(df['sex'])

# Print the original and transformed data
print("The original data")
print(original_sex)
print("The transformed data using LabelEncoder")
print(df['sex'].values)

The original data
0       0
1       1
2       1
3       1
4       1
       ..
1333    1
1334    0
1335    0
1336    0
1337    0
Name: sex, Length: 1338, dtype: int64
The transformed data using LabelEncoder
[0 1 1 ... 0 0 0]


Modify the dataframe to replace the 'sex' column with the transformed data,

In [ ]:
df

age  sex     bmi  children smoker     region      charges
0      19    0  27.900         0    yes  southwest  16884.92400
1      18    1  33.770         1     no  southeast   1725.55230
2      28    1  33.000         3     no  southeast   4449.46200
3      33    1  22.705         0     no  northwest  21984.47061
4      32    1  28.880         0     no  northwest   3866.85520
...   ...  ...     ...       ...    ...        ...          ...
1333   50    1  30.970         3     no  northwest  10600.54830
1334   18    0  31.920         0     no  northeast   2205.98080
1335   18    0  36.850         0     no  southeast   1629.83350
1336   21    0  25.800         0     no  southwest   2007.94500
1337   61    0  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

 Which number is assigned for male and female?

In [ ]:
print(labelenc.classes_)

[0 1]


female is 0 and male is 1

DATAPREPROCESSING: Convert 'smoker' to a numerical variable through label encoder

In [ ]:
# Copy the original 'smoker' column
original_smoker = df['smoker'].copy()

# Initialize the LabelEncoder
labelenc = LabelEncoder()

# Apply LabelEncoder to the 'smoker' column
df['smoker'] = labelenc.fit_transform(df['smoker'])

# Print the original and transformed data
print("The original data")
print(original_smoker)
print("The transformed data using LabelEncoder")
print(df['smoker'].values)

The original data
0       yes
1        no
2        no
3        no
4        no
       ... 
1333     no
1334     no
1335     no
1336     no
1337    yes
Name: smoker, Length: 1338, dtype: object
The transformed data using LabelEncoder
[1 0 0 ... 0 0 1]


Modify the dataframe to replace the 'smoker' column with the transformed data,

In [ ]:
df

age  sex     bmi  children  smoker     region      charges
0      19    0  27.900         0       1  southwest  16884.92400
1      18    1  33.770         1       0  southeast   1725.55230
2      28    1  33.000         3       0  southeast   4449.46200
3      33    1  22.705         0       0  northwest  21984.47061
4      32    1  28.880         0       0  northwest   3866.85520
...   ...  ...     ...       ...     ...        ...          ...
1333   50    1  30.970         3       0  northwest  10600.54830
1334   18    0  31.920         0       0  northeast   2205.98080
1335   18    0  36.850         0       0  southeast   1629.83350
1336   21    0  25.800         0       0  southwest   2007.94500
1337   61    0  29.070         0       1  northwest  29141.36030

[1338 rows x 7 columns]

Which number is assigned for yes and no? no=0 yes=1

In [ ]:
print(labelenc.classes_)

['no' 'yes']


DATAPREPROCESSING: Convert 'region' to a numerical variable through label encoder

In [ ]:
# Copy the original 'region' column
original_region = df['region'].copy()

# Initialize the LabelEncoder
labelenc = LabelEncoder()

# Apply LabelEncoder to the 'region' column
df['region'] = labelenc.fit_transform(df['region'])

# Print the original and transformed data
print("The original data")
print(original_region)
print("The transformed data using LabelEncoder")
print(df['region'].values)

The original data
0       southwest
1       southeast
2       southeast
3       northwest
4       northwest
          ...    
1333    northwest
1334    northeast
1335    southeast
1336    southwest
1337    northwest
Name: region, Length: 1338, dtype: object
The transformed data using LabelEncoder
[3 2 2 ... 2 3 1]


Modify the dataframe to replace the 'region' column with the transformed data,

In [ ]:
df

age  sex     bmi  children  smoker  region      charges
0      19    0  27.900         0       1       3  16884.92400
1      18    1  33.770         1       0       2   1725.55230
2      28    1  33.000         3       0       2   4449.46200
3      33    1  22.705         0       0       1  21984.47061
4      32    1  28.880         0       0       1   3866.85520
...   ...  ...     ...       ...     ...     ...          ...
1333   50    1  30.970         3       0       1  10600.54830
1334   18    0  31.920         0       0       0   2205.98080
1335   18    0  36.850         0       0       2   1629.83350
1336   21    0  25.800         0       0       3   2007.94500
1337   61    0  29.070         0       1       1  29141.36030

[1338 rows x 7 columns]

Which numbers are assigned for each regions?

0 northeast
1 northwest
2 southeast
3 southwest

In [ ]:
print(labelenc.classes_)

['northeast' 'northwest' 'southeast' 'southwest']


#3. Feature Selection + Linear Regression

Run a multiple linear regression machine learning algorithm which IVs are selected through SelectKbest f_regression. Select the best 3.

In [ ]:
# Calculate correlations between all columns (this computes the Pearson correlation coefficients between all pairs of numeric columns in the DataFrame)
correlations = df.corr()

# Identify potential features based on correlation with other columns
# Exclude highly correlated columns and non-numeric columns like 'ID'

# Filter out non-numeric columns and excluded columns
numeric_columns = df.select_dtypes(include=[np.number]).columns # selects only numeric columns
# filtered_features = [col for col in numeric_columns if col not in exclude_columns]
filtered_features = list(numeric_columns)

# Determine the target variable
# Find the column with the highest average correlation with other columns
avg_correlations = correlations.abs().mean() # computes mean of absolute correlation values for each column
target_variable = avg_correlations.idxmax() # identifies column w/highest average correlation (idxmax - index max)

# Define the feature matrix X and target variable y
X = df[filtered_features] # creates matrix x w/identified feature colulmns
y = df[target_variable] # sets target y to identified target column

# Display the identified features and target variable
print("\nIdentified Features:")
print(filtered_features)
print("\nTarget Variable:")
print(target_variable)


Identified Features:
['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges']

Target Variable:
charges


Target: charges

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error

X = df.drop(['charges'], axis=1)
y = df['charges']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42)

In [ ]:
df

age  sex     bmi  children  smoker  region      charges
0      19    0  27.900         0       1       3  16884.92400
1      18    1  33.770         1       0       2   1725.55230
2      28    1  33.000         3       0       2   4449.46200
3      33    1  22.705         0       0       1  21984.47061
4      32    1  28.880         0       0       1   3866.85520
...   ...  ...     ...       ...     ...     ...          ...
1333   50    1  30.970         3       0       1  10600.54830
1334   18    0  31.920         0       0       0   2205.98080
1335   18    0  36.850         0       0       2   1629.83350
1336   21    0  25.800         0       0       3   2007.94500
1337   61    0  29.070         0       1       1  29141.36030

[1338 rows x 7 columns]

In [ ]:
# Create a SelectKBest object with the f_regression scoring function
selector = SelectKBest(score_func=f_regression, k=3) #k=3 select 3

# Fit the selector to the training data and transform the features
X_train_selected = selector.fit_transform(X_train, y_train)
#Instead of having all features, X train and y train will not just have the best 5 columns

# Apply the same transformation to the test data
X_test_selected = selector.transform(X_test)
#y_test is not needed since we only use it in checking errors.

# Train a linear regression model using the selected features
model = LinearRegression()
model.fit(X_train_selected, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test_selected)

y_pred

array([ 8126.0381885 ,  7394.28272886, 37158.84838683,  8581.0070836 ,
       27030.36777024, 11211.03803528,   396.46090003, 17411.01188025,
        1000.10342267, 10362.28081412, 27385.87456047,  8900.44958234,
        5409.7163993 , 37642.31697061, 40839.40061467, 36755.4714203 ,
       15078.85973148, 35976.67961908,  9358.11318828, 31640.05910532,
        2440.19577903, 10103.69857039,  3054.80434273,  7013.99800221,
       10432.91089822, 13021.46410057, 15414.13816324,  5157.2002595 ,
        9132.06068144,  2704.34863308,  8202.74038557, 13521.60846112,
        5087.41602819,  2422.49296085,  4759.54504675, 13058.38626292,
        2231.84609037,  9193.42243909, 33076.43571845, 32603.77602325,
        2447.95358853,  4163.48515401, 14598.93772038, 11901.3157456 ,
        8537.69249428, 12589.07993369,  4669.3632864 ,  2575.05150218,
       34702.96971749,  8902.14128791, 16607.10380149,  2444.75136601,
       12288.86879718,  1350.73431191, 14137.05686857, 11778.09673669,
      

In [ ]:
# Print the indices of the selected features
selected_feature_indices = selector.get_support(indices=True)

# Get the indices of the selected features
print(f"Selected feature indices: {selected_feature_indices}")

Selected feature indices: [0 2 4]


In [ ]:
# Get the names of the selected features from the original DataFrame
selected_feature_names = X.columns[selected_feature_indices]

# Print the names of the selected features
print(f"Selected feature names: {selected_feature_names}")


Selected feature names: Index(['age', 'bmi', 'smoker'], dtype='object')


#4. Evaluation

What is the MAPE, MAE, and RMSE? Check for both train and test data.


In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

# Function to calculate Mean Absolute Percentage Error (MAPE)
def mean_absolute_percentage_error(y_true, y_pred):
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

# Train set evaluation
y_train_pred = model.predict(X_train_selected)

mape_train = mean_absolute_percentage_error(y_train, y_train_pred)
mae_train = mean_absolute_error(y_train, y_train_pred)
rmse_train = np.sqrt(mean_squared_error(y_train, y_train_pred))

print("Training Set Metrics with Selected Features:")
print("MAPE:", mape_train)
print("MAE:", mae_train)
print("RMSE:", rmse_train)

# Test set evaluation
y_test_pred = model.predict(X_test_selected)

mape_test = mean_absolute_percentage_error(y_test, y_test_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)
rmse_test = np.sqrt(mean_squared_error(y_test, y_test_pred))

print("\nTest Set Metrics with Selected Features:")
print("MAPE:", mape_test)
print("MAE:", mae_test)
print("RMSE:", rmse_test)

Training Set Metrics with Selected Features:
MAPE: 43.58243570372571
MAE: 4256.460560635411
RMSE: 6144.0245425254025

Test Set Metrics with Selected Features:
MAPE: 46.91166555808011
MAE: 4274.4319736848465
RMSE: 5997.788247994093


**Train Set:**

MAPE: Percentage of error at 43.58% in relation to actual values.

MAE: Average prediction error 4256.46 means predictions are off from actual by ~ 4256.46 dollars.

RMSE: Root of average square prediction error at 6,144.02 which relects the differences between predicted and actual value in units (charge dollars).

**Test Set:**

MAPE: Percentage of error at 46.91% relative to the actual values and is a little higher than the train set.

MAE: Average prediction error at 4274.43 means predictions are off from actual by ~4,274.43 dollars.

RMSE: Root of average square predition error at 5,997.79 which relects the differences between predicted and actual value in units (charge dollars).



#5. Verdict
Did it perform well on the training and test phase?

The model does show consistency between the training and test sets and performs somewhat well in predicting the charges. There is room for improvement for accuracy and reduction of the percentage error in MAPE.

**Save file for future use.**

In [ ]:
from joblib import dump

# Assuming 'model' is your trained LinearRegression model
dump(model, 'trained_model.joblib')



['trained_model.joblib']